Add alpha_vantage configurable scan interval

matthiasbeyer opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This makes the "scan interval" of the `alpha_vantage` integration configurable (was 5 minutes before).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #109153

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

matthiasbeyer committed:
alpha_vantage: Add configurable scan interval

Signed-off-by: Matthias Beyer <mail@beyermatthias.de>

joostlek reviewed (changes requested):
Users can change their polling interval by disabling polling and making use of the `homeassistant.update_entity` service. This gives way more flexibility in how you want to poll.

Thus we don't allow integrations to make their own polling interval configurable. So I have to decline your PR. Nevertheless, thanks for the willingness to contribute

matthiasbeyer commented:
> Users can change their polling interval by disabling polling and making use of the homeassistant.update_entity service. This gives way more flexibility in how you want to poll.

Can you point me to the relevant documentation how to disable sensor polling from the `configuration.yaml`? I cannot find it with googling around...

That would close the linked issue then, I guess?

joostlek commented:
Oh wait, that's a good point lol. That only works for config flow. My bad.

Can you maybe check what the default platform scheme is? I can vaguely remember that it already had one of them

matthiasbeyer commented:
> Can you maybe check what the default platform scheme is? I can vaguely remember that it already had one of them

I am 100% new to HA (literally started my installation 2 days ago), so I don't even know what that means... but if you point me to things, I am eager to learn!

joostlek commented:
Oh, in your PR you extend the platform schema, can you check the base implementation? I think it already has something similar, but not 100% sure. (I'm on mobile so I can't check)

matthiasbeyer commented:
Ah, I see. Yes, the `PLATFORM_SCHEMA` indeed has a `CONF_SCAN_INTERVAL`, which is optional though.
If you want I can modify my PR to keep the interval optional for the component!

joostlek commented:
I think that is the platform equivalent of overwriting the scan interval, so maybe try that first

matthiasbeyer committed:
alpha_vantage: Remove overwrite for CONF_SCAN_INTERVAL

Apparently, the platform schema already has that setting, so we dont
need to overwrite it here again.

Signed-off-by: Matthias Beyer <mail@beyermatthias.de>

MartinHjelmare commented on homeassistant/components/alpha_vantage/sensor.py:
```diff
@@ -79,9 +77,13 @@ def setup_platform(
     conversions: list[dict[str, str]] = config.get(CONF_FOREIGN_EXCHANGE, [])
 
+    _LOGGER.debug("Scan interval = %s", scan_interval)
+    self.SCAN_INTERVAL = timedelta(minutes=scan_interval)
```
This doesn't work. The scan interval can already be changed in the platform schema as pointed out in the discussion above.
